src/math.jl

```diff
@@ -1,41 +1,60 @@
-Base.:*(l::AbstractDimensions, r::AbstractDimensions) = map_dimensions(+, l, r)
-Base.:*(l::AbstractQuantity, r::AbstractQuantity) = new_quantity(typeof(l), ustrip(l) * ustrip(r), dimension(l) * dimension(r))
-Base.:*(l::AbstractQuantity, r::AbstractDimensions) = new_quantity(typeof(l), ustrip(l), dimension(l) * r)
-Base.:*(l::AbstractDimensions, r::AbstractQuantity) = new_quantity(typeof(r), ustrip(r), l * dimension(r))
-Base.:*(l::AbstractQuantity, r) = new_quantity(typeof(l), ustrip(l) * r, dimension(l))
-Base.:*(l, r::AbstractQuantity) = new_quantity(typeof(r), l * ustrip(r), dimension(r))
-Base.:*(l::AbstractDimensions, r) = error("Please use an `AbstractQuantity` for multiplication. You used multiplication on types: $(typeof(l)) and $(typeof(r)).")
-Base.:*(l, r::AbstractDimensions) = error("Please use an `AbstractQuantity` for multiplication. You used multiplication on types: $(typeof(l)) and $(typeof(r)).")
+for (type, base_type) in ABSTRACT_QUANTITY_TYPES
+    @eval begin
+        Base.:*(l::$type, r::$type) = new_quantity(typeof(l), ustrip(l) * ustrip(r), dimension(l) * dimension(r))
+        Base.:/(l::$type, r::$type) = new_quantity(typeof(l), ustrip(l) / ustrip(r), dimension(l) / dimension(r))
 
-Base.:/(l::AbstractDimensions, r::AbstractDimensions) = map_dimensions(-, l, r)
-Base.:/(l::AbstractQuantity, r::AbstractQuantity) = new_quantity(typeof(l), ustrip(l) / ustrip(r), dimension(l) / dimension(r))
-Base.:/(l::AbstractQuantity, r::AbstractDimensions) = new_quantity(typeof(l), ustrip(l), dimension(l) / r)
-Base.:/(l::AbstractDimensions, r::AbstractQuantity) = new_quantity(typeof(r), inv(ustrip(r)), l / dimension(r))
-Base.:/(l::AbstractQuantity, r) = new_quantity(typeof(l), ustrip(l) / r, dimension(l))
-Base.:/(l, r::AbstractQuantity) = l * inv(r)
-Base.:/(l::AbstractDimensions, r) = error("Please use an `AbstractQuantity` for division. You used division on types: $(typeof(l)) and $(typeof(r)).")
-Base.:/(l, r::AbstractDimensions) = error("Please use an `AbstractQuantity` for division. You used division on types: $(typeof(l)) and $(typeof(r)).")
-
-Base.:+(l::AbstractQuantity, r::AbstractQuantity) =
-    let
-        dimension(l) == dimension(r) || throw(DimensionError(l, r))
-        new_quantity(typeof(l), ustrip(l) + ustrip(r), dimension(l))
-    end
-Base.:-(l::AbstractQuantity) = new_quantity(typeof(l), -ustrip(l), dimension(l))
-Base.:-(l::AbstractQuantity, r::AbstractQuantity) = l + (-r)
+        Base.:*(l::$type, r::$base_type) = new_quantity(typeof(l), ustrip(l) * r, dimension(l))
+        Base.:/(l::$type, r::$base_type) = new_quantity(typeof(l), ustrip(l) / r, dimension(l))
+
+        Base.:*(l::$base_type, r::$type) = new_quantity(typeof(r), l * ustrip(r), dimension(r))
+        Base.:/(l::$base_type, r::$type) = new_quantity(typeof(r), l / ustrip(r), inv(dimension(r)))
+
+        Base.:*(l::$type, r::AbstractDimensions) = new_quantity(typeof(l), ustrip(l), dimension(l) * r)
```

Comment:
Hm, I don't quite follow why we need to modify the `AbstractDimensions` methods to e.g. dispatch the other argument on `$type`, rather than leaving it as before. I also don't quite understand the purpose of the new methods in `disambiguities.jl`. Naively I'd hope that we could just leave all the `AbstractDimensions` methods alone, but perhaps I'm missing something?

Reply:
I thought so too. But `Aqua` kept detecting various method ambiguities. A lot of the complexity in this PR is purely for the reason of eliminating different ambiguities. (Some of those are new in 1.10 so may not be visible in older versions)

Reply:
Part of the issue is because `AbstractUnionQuantity` is on par with `Number` in a type hierarchy. So whenever `Number` appears in a method, I have to break the union into each type.

Reply:
I did expect the ambiguities with `*(::Number, ::Number)`, those unfortunately seem avoidable when trying to make a new number type participate in the type hieracy. Here's an example of a similar flavor in `ForwardDiff` (`Dual` overloads `Real`, for reference): https://github.com/JuliaDiff/ForwardDiff.jl/blob/d3002093beb88ff0b98ed178377961dfd55c1247/src/dual.jl#L142C14-L148.

I'm mostly confused about the `AbstractDimensions` methods, rather than the `AbstractQuantity` ones. Since we have total control over how `AbstractDimensions` is used, I was surprised that the dimensions methods needed to be included in the metaprogramming loops, rather than kept outside.

Reply:
To put it more simply: it seems to me that, in theory, this PR should not affect `AbstractDimensions` at all, e.g. it doesn't change its position in the type hierarchy. So why should the methods defined on it change?

Reply:
Oh I think I understand your question now. The answer lies two steps away. First: there's a method:
```julia
Base.:*(l::AbstractDimensions, r::Number)
```
(whose only job is to raise a helpful error).

In this PR, before the ambiguities, I had defined

```julia
Base.:*(l::AbstractDimensions, r::AbstractUnionQuantity)
```

But because `AbstractUnionQuantity` and `Number` are at the same level of the hierarchy (one above `AbstractQuantity`), I therefore had to split the `AbstractUnionQuantity` into

```julia
Base.:*(l::AbstractDimensions, r::AbstractQuantity)
Base.:*(l::AbstractDimensions, r::AbstractGenericQuantity)
```

to get rid of the ambiguity.